outlines/caching.py

```diff
@@ -1,20 +1,79 @@
+import asyncio
+import hashlib
 import os
 from typing import Callable, Optional
 
-from perscache import Cache, NoCache
-from perscache.serializers import JSONSerializer
-from perscache.storage import LocalFileStorage
+import cloudpickle
+import torch
+from diskcache import Cache
 
 home_dir = os.path.expanduser("~")
 cache_dir = os.environ.get("OUTLINES_CACHE_DIR", f"{home_dir}/.cache/outlines")
-memory = Cache(serializer=JSONSerializer(), storage=LocalFileStorage(cache_dir))
+memory = Cache(cache_dir, eviction_policy="none", cull_limit=0)
+_caching_enabled = True
 
 
-def cache(ignore: Optional[str] = None):
-    def cache_fn(fn: Callable):
-        return memory.cache(ignore=ignore)(fn)
+def hash_data(*data) -> str:
+    """Pickles and hashes all the data passed to it as args.
+    Pickling and then hashing significantly reduces the size of the key especially when dealing with large tensors.
+    """
+    result = hashlib.md5()  # nosec B303
+    for datum in data:
+        if isinstance(datum, torch.Tensor):
+            datum = datum.cpu().numpy()
+        result.update(cloudpickle.dumps(datum))
```

Comment:
IMHO, since you're already defining a key function for each hashable, the key function should be responsible for ensuring hashability, rather than ensuring arbitrary objects are dumpable via `cloudpickle`, which may have non-deterministic outputs: https://github.com/cloudpipe/cloudpickle/issues/385

Reply:
I removed the line regarding `torch.Tensor` that I had picked up from #441 as it's not needed for the issue this PR tries to solve. On hashability, I wanted to give the option not to provide a key function if we just want to use the arguments of the cached function for the key (the standard situation I guess) so that we don't have to encumber the file of the cached function with an extra function definition.